### DISCUSSION TITLE
Adding a property to an existing pset (without using the api)

### DISCUSSION THREAD

**Discussion by DaivaMarcinkeviciute** [IfcOpenShell/IfcOpenShell#3786] [General]
I'm writing a code that adds a property "IsExternal" to the Pset "Pset_SpaceCommon", in case it's not present (value: False). The code works fine except for the last line. I cannot find a way to append the property without installing the ifcopenshell.api. Could someone help me?

```
for space in ifcFile.by_type("IfcSpace"):
    # Create a new property set "Pset_SpaceCommon" if it doesn't exist
    pset_spaceCommon = None
    for pset in space.IsDefinedBy:
        if pset.RelatingPropertyDefinition.Name == "Pset_SpaceCommon":
            pset_spaceCommon = pset.RelatingPropertyDefinition
            break
    if pset_spaceCommon is None:
        pset_spaceCommon = ifcFile.createIfcPropertySet("Pset_SpaceCommon", "Space Common Properties")
        space.IsDefinedBy.append(ifcFile.createIfcRelDefinesByProperties(None, None, [pset_spaceCommon]))

    # Check if "IsExternal" property already exists in the property set
    is_external_property = None
    for prop in pset_spaceCommon.HasProperties:           
        if prop.Name == "IsExternal":
            prop.NominalValue.wrappedValue = False
            is_external_property = prop
            break

    # If "IsExternal" property doesn't exist, create and add it to the property set
    if is_external_property is None:
        is_external_property = ifcFile.createIfcPropertySingleValue("IsExternal", "IsExternal", ifcFile.create_entity("IfcBoolean", False), None)
        pset_spaceCommon.HasProperties.append(is_external_property) #This line needs to be replaced
```

**Comment by aothms** [comment]
Yeah, this is one of last remaining big inconveniences in the API.

        pset_spaceCommon.HasProperties += (is_external_property,)

You need to actually assign for the file to update. Just appending only updates a local copy that never makes it back to the model.

**Reply by DaivaMarcinkeviciute** [reply]
Thank You! Could you please also explain how to assign the file to update?

**Reply by aothms** [reply]
ifcFile.write(fn)

**Reply by DaivaMarcinkeviciute** [reply]
I did notice that the newly added property seams to disapear as soon as i move into the next function and call the ifcSpace instance again.
But how do i make it stay exactly?

Doesn't ifcFile.write(fn) simpy write a new ifc file? (fn is the path i assume?)
Do i need to write a new ifc file everytime i add a property to a pset and then read it again?

**Reply by aothms** [reply]
Yes, IFC-SPF is a text file. There isn't a general approach that would make partial updates significantly more efficient.
